File: mobile_money_gateway/utils/ussd_executor.py
from typing import Optional, List, Dict
from enum import Enum


class USSDNetwork(str, Enum):
    MTN = "MTN"
    AIRTEL = "Airtel"
    ZAMTEL = "Zamtel"


class USSDOperation(str, Enum):
    CASH_IN = "cash_in"
    CASH_OUT = "cash_out"
    BALANCE_CHECK = "balance_check"
    MINI_STATEMENT = "mini_statement"


class USSDCommand:
    def __init__(self, network: USSDNetwork, operation: USSDOperation, slot_index: str):
        self.network = network
        self.operation = operation
        self.slot_index = slot_index
        self.steps: List[Dict] = []

    def build(self, phone_number: str, amount: float, pin: Optional[str] = None) -> List[Dict]:
        if self.network == USSDNetwork.MTN:
            return self._build_mtn(phone_number, amount, pin)
        elif self.network == USSDNetwork.AIRTEL:
            return self._build_airtel(phone_number, amount, pin)
        elif self.network == USSDNetwork.ZAMTEL:
            return self._build_zamtel(phone_number, amount, pin)
        return []

    def _build_mtn(self, phone_number: str, amount: float, pin: Optional[str] = None) -> List[Dict]:
        steps = []
        if self.operation == USSDOperation.CASH_OUT:
            steps.extend([
                {"type": "ussd", "code": "*111#", "sim_slot": self.slot_index},
                {"type": "ussd_response", "expected_pattern": r"(?:Send Money|Transfer)"},
                {"type": "ussd_input", "value": "1"},
                {"type": "ussd_response", "expected_pattern": r"(?:Enter.*number|Phone)"},
                {"type": "ussd_input", "value": phone_number},
                {"type": "ussd_response", "expected_pattern": r"(?:amount|Enter)"},
                {"type": "ussd_input", "value": str(int(amount))},
                {"type": "ussd_response", "expected_pattern": r"(?:PIN|password)"},
                {"type": "ussd_input", "value": pin or ""},
            ])
        elif self.operation == USSDOperation.CASH_IN:
            steps.extend([
                {"type": "ussd", "code": "*112#", "sim_slot": self.slot_index},
                {"type": "ussd_response", "expected_pattern": r"(?:Receive Money|Deposit)"},
                {"type": "ussd_input", "value": "1"},
                {"type": "ussd_response", "expected_pattern": r"(?:Enter.*number|Phone)"},
                {"type": "ussd_input", "value": phone_number},
                {"type": "ussd_response", "expected_pattern": r"(?:amount|Enter)"},
                {"type": "ussd_input", "value": str(int(amount))},
            ])
        return steps

    def _build_airtel(self, phone_number: str, amount: float, pin: Optional[str] = None) -> List[Dict]:
        steps = []
        if self.operation == USSDOperation.CASH_OUT:
            steps.extend([
                {"type": "ussd", "code": "*115#", "sim_slot": self.slot_index},
                {"type": "ussd_response", "expected_pattern": r"(?:Send Money|Transfer)"},
                {"type": "ussd_input", "value": "1"},
                {"type": "ussd_response", "expected_pattern": r"(?:Enter.*number|Phone)"},
                {"type": "ussd_input", "value": phone_number},
                {"type": "ussd_response", "expected_pattern": r"(?:amount|Enter)"},
                {"type": "ussd_input", "value": str(int(amount))},
                {"type": "ussd_response", "expected_pattern": r"(?:PIN|password)"},
                {"type": "ussd_input", "value": pin or ""},
            ])
        elif self.operation == USSDOperation.CASH_IN:
            steps.extend([
                {"type": "ussd", "code": "*114#", "sim_slot": self.slot_index},
                {"type": "ussd_response", "expected_pattern": r"(?:Receive Money|Deposit)"},
                {"type": "ussd_input", "value": "1"},
                {"type": "ussd_response", "expected_pattern": r"(?:Enter.*number|Phone)"},
                {"type": "ussd_input", "value": phone_number},
                {"type": "ussd_response", "expected_pattern": r"(?:amount|Enter)"},
                {"type": "ussd_input", "value": str(int(amount))},
            ])
        return steps

    def _build_zamtel(self, phone_number: str, amount: float, pin: Optional[str] = None) -> List[Dict]:
        steps = []
        if self.operation == USSDOperation.CASH_OUT:
            steps.extend([
                {"type": "ussd", "code": "*811#", "sim_slot": self.slot_index},
                {"type": "ussd_response", "expected_pattern": r"(?:Send Money|Transfer)"},
                {"type": "ussd_input", "value": "1"},
                {"type": "ussd_response", "expected_pattern": r"(?:Enter.*number|Phone)"},
                {"type": "ussd_input", "value": phone_number},
                {"type": "ussd_response", "expected_pattern": r"(?:amount|Enter)"},
                {"type": "ussd_input", "value": str(int(amount))},
                {"type": "ussd_response", "expected_pattern": r"(?:PIN|password)"},
                {"type": "ussd_input", "value": pin or ""},
            ])
        elif self.operation == USSDOperation.CASH_IN:
            steps.extend([
                {"type": "ussd", "code": "*810#", "sim_slot": self.slot_index},
                {"type": "ussd_response", "expected_pattern": r"(?:Receive Money|Deposit)"},
                {"type": "ussd_input", "value": "1"},
                {"type": "ussd_response", "expected_pattern": r"(?:Enter.*number|Phone)"},
                {"type": "ussd_input", "value": phone_number},
                {"type": "ussd_response", "expected_pattern": r"(?:amount|Enter)"},
                {"type": "ussd_input", "value": str(int(amount))},
            ])
        return steps

    def to_dict(self) -> Dict:
        return {
            "network": self.network.value,
            "operation": self.operation.value,
            "slot_index": self.slot_index,
            "steps": self.steps,
        }

File: GatewayApp/app/src/main/java/com/mobilemoney/gateway/util/USSDExecutor.kt
package com.mobilemoney.gateway.util

import android.util.Log
import com.mobilemoney.gateway.model.USSDStep
import java.util.regex.Pattern

object USSDExecutor {
    private const val TAG = "USSDExecutor"

    fun buildMTN(operation: String, slotIndex: String, phoneNumber: String, amount: Double, pin: String?): List<USSDStep> {
        return when {
            operation.equals("balance", true) -> mutableListOf(
                // MTN Zambia balance check: dial *115# → single-step, response contains balance
                USSDStep("ussd", "*115#", null, slotIndex),
                USSDStep("ussd_response", null, "(?i)balance|kwacha|ZMW|K\\s*[\\d,]+|your.*balance|MoMo)")
            )
            operation.equals("cash_out", true) -> mutableListOf(
                USSDStep("ussd", "*115#", null, slotIndex),
                USSDStep("ussd_response", null, "(?:Send Money|Transfer|menu)"),
                USSDStep("ussd_input", "1"),
                USSDStep("ussd_response", null, "(?:Enter.*number|Phone|recipient)"),
                USSDStep("ussd_input", phoneNumber),
                USSDStep("ussd_response", null, "(?:amount|Enter|how much)"),
                USSDStep("ussd_input", amount.toInt().toString()),
                USSDStep("ussd_response", null, "(?:PIN|password|secret)"),
                USSDStep("ussd_input", pin.orEmpty())
            )
            else -> mutableListOf(
                // cash_in / deposit
                USSDStep("ussd", "*115#", null, slotIndex),
                USSDStep("ussd_response", null, "(?:Receive Money|Deposit|menu)"),
                USSDStep("ussd_input", "5"),
                USSDStep("ussd_response", null, "(?:Enter.*number|Phone|agent)"),
                USSDStep("ussd_input", phoneNumber),
                USSDStep("ussd_response", null, "(?:amount|Enter|how much)"),
                USSDStep("ussd_input", amount.toInt().toString())
            )
        }
    }

    fun buildAirtel(operation: String, slotIndex: String, phoneNumber: String, amount: Double, pin: String?): List<USSDStep> {
        return if (operation.equals("cash_out", true)) {
            mutableListOf(
                USSDStep("ussd", "*115#", null, slotIndex),
                USSDStep("ussd_response", null, "(?:Send Money|Transfer)"),
                USSDStep("ussd_input", "1"),
                USSDStep("ussd_response", null, "(?:Enter.*number|Phone)"),
                USSDStep("ussd_input", phoneNumber),
                USSDStep("ussd_response", null, "(?:amount|Enter)"),
                USSDStep("ussd_input", amount.toInt().toString()),
                USSDStep("ussd_response", null, "(?:PIN|password)"),
                USSDStep("ussd_input", pin.orEmpty())
            )
        } else {
            mutableListOf(
                USSDStep("ussd", "*114#", null, slotIndex),
                USSDStep("ussd_response", null, "(?:Receive Money|Deposit)"),
                USSDStep("ussd_input", "1"),
                USSDStep("ussd_response", null, "(?:Enter.*number|Phone)"),
                USSDStep("ussd_input", phoneNumber),
                USSDStep("ussd_response", null, "(?:amount|Enter)"),
                USSDStep("ussd_input", amount.toInt().toString())
            )
        }
    }

    fun buildZamtel(operation: String, slotIndex: String, phoneNumber: String, amount: Double, pin: String?): List<USSDStep> {
        return if (operation.equals("cash_out", true)) {
            mutableListOf(
                USSDStep("ussd", "*811#", null, slotIndex),
                USSDStep("ussd_response", null, "(?:Send Money|Transfer)"),
                USSDStep("ussd_input", "1"),
                USSDStep("ussd_response", null, "(?:Enter.*number|Phone)"),
                USSDStep("ussd_input", phoneNumber),
                USSDStep("ussd_response", null, "(?:amount|Enter)"),
                USSDStep("ussd_input", amount.toInt().toString()),
                USSDStep("ussd_response", null, "(?:PIN|password)"),
                USSDStep("ussd_input", pin.orEmpty())
            )
        } else {
            mutableListOf(
                USSDStep("ussd", "*810#", null, slotIndex),
                USSDStep("ussd_response", null, "(?:Receive Money|Deposit)"),
                USSDStep("ussd_input", "1"),
                USSDStep("ussd_response", null, "(?:Enter.*number|Phone)"),
                USSDStep("ussd_input", phoneNumber),
                USSDStep("ussd_response", null, "(?:amount|Enter)"),
                USSDStep("ussd_input", amount.toInt().toString())
            )
        }
    }

    fun matchesPattern(text: String?, expectedPattern: String?): Boolean {
        if (text.isNullOrBlank() || expectedPattern.isNullOrBlank()) return false
        return try {
            Pattern.compile(expectedPattern, Pattern.CASE_INSENSITIVE).matcher(text).find()
        } catch (e: Exception) {
            Log.w(TAG, "Invalid regex pattern: $expectedPattern", e)
            false
        }
    }
}
